src/test/java/io/confluent/examples/streams/CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,429 @@
+          sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(timestamp);
+        }
+
+        private void sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(final long timestamp) {
+          try (KeyValueIterator<String, Pair<Double, Instant>> iterator = streamBufferStore.all()) {
+            while (iterator.hasNext()) {
+              final KeyValue<String, Pair<Double, Instant>> record = iterator.next();
+              final Duration delta = Duration.between(record.value.y, Instant.ofEpochMilli(timestamp));
+              if (delta.compareTo(maxWaitTimePerRecordForTableSideData) > 0) {
+                // Final attempt to fetch table-side data; we use that data even if it is null (indicates: missing).
+                final Long tableValue = tableStore.get(record.key);
+                streamBufferStore.delete(record.key);
+                context.forward(record.key, new Pair<>(record.value.x, tableValue));
+              }
+            }
+          }
+        }
+
+        @Override
+        public void close() {
+        }
+
+      };
+    }
+
+  }
+
+  /**
+   * Implements table-side triggering of join output.
+   *
+   * For every <i>observed </i> record arriving at its upstream table, this transformer will check for a buffered (i.e.,
+   * not yet joined) record on the stream side.  If there is a match, then the transformer will produce a fully
+   * populated join output message -- which is the desired table-side triggering behavior.  If there is no match, then
+   * the transformer will do nothing.
+   */
+  private static final class StreamTableJoinTableSideTrigger
+    implements ValueTransformerWithKeySupplier<String, Long, Pair<Double, Long>> {
+
+    final private String streamBufferStoreName;
+
+    StreamTableJoinTableSideTrigger(final String streamBufferStoreName) {
+      this.streamBufferStoreName = streamBufferStoreName;
+    }
+
+    @Override
+    public ValueTransformerWithKey<String, Long, Pair<Double, Long>> get() {
+      return new ValueTransformerWithKey<String, Long, Pair<Double, Long>>() {
+
+        private KeyValueStore<String, Pair<Double, Instant>> streamBufferStore;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+          streamBufferStore = (KeyValueStore<String, Pair<Double, Instant>>) context.getStateStore(streamBufferStoreName);
+        }
+
+        @Override
+        public Pair<Double, Long> transform(final String key, final Long value) {
+          return possiblySendFullJoinRecord(key, value);
+        }
+
+        private Pair<Double, Long> possiblySendFullJoinRecord(final String key, final Long value) {
+          if (value != null) {
+            final Pair<Double, Instant> streamValue = streamBufferStore.get(key);
+            if (streamValue != null) {
+              // We have data from both stream and table, so we can send a fully populated join message downstream.
```

Comment:
This does not consider event-time. We should only join, if `context.timestamp()` (that is the table-update-record-timestamp) is smaller than or equal to the stream-side-record's timestamp (that is available in the provided `Pair`).

Furthermore, if the table-update-record's timestamp is larger than the stream-side-record, we know that we can emit the stream-side record as left-join result with `null`-table-part (and also delete it from the store).

Reply:
You are right, of course.  But this example intentionally changes the join semantics to something custom, and it's up to the user and the use case to decide whether or not this behavior is wanted.  However, I get your point that we must be more explicit what the goal and behavior of this example is.  I updated the example docs accordingly.

As a result of thinking through this and updating the example, I also spotted a bug that I fixed along the way: when a new stream record arrives, then any buffered stream record for the same key must be force-forwarded downstream.

Reply:
I think it would be helpful to users to add this exchange (massaged a bit of course) into the Javadoc unless it's already there and I've missed it.

Reply:
@bbejeck: Could you re-read the top-level example javadocs again (for the test file)?  I had already reworded the text after the exchange with @mjsax.